### ISSUE TITLE
Verification Error: is null input must not be null in LoopUnswitchingPhase

### ISSUE THREAD

**Issue by smarr** [graalvm/graal-core#244]
After updating to d1057cb806157581b47ed071cf7a73ccc63dc59b (of Jan 14th, I update about once a month).

I was not able to reproduce on my local machine. I doubt the macOS makes a differences. It is more likely that the timing on my Linux benchmark machine is different and more likely to compile the problematic method/graph.

I am seeing sporadic benchmark failures with the following error:

```
[thread:25] scope: TruffleCompilerThread-25
  [thread:25] scope: TruffleCompilerThread-25.Truffle
  Context: Truffle<Method_Vacation>>#initializeManager:	@330eef1()>
    [thread:25] scope: TruffleCompilerThread-25.Truffle.TruffleGraal.GraalCompiler
    Context: StructuredGraph:33372{Method Vacation>>#initializeManager:	@330eef1, HotSpotMethod<OptimizedCallTarget.callRoot(Object[])>}
        [thread:25] scope: TruffleCompilerThread-25.Truffle.TruffleGraal.GraalCompiler.GraalCompilerRoot.GraalCompiler
        Context: StructuredGraph:33372{Method Vacation>>#initializeManager:	@330eef1, HotSpotMethod<OptimizedCallTarget.callRoot(Object[])>}
                [thread:25] scope: TruffleCompilerThread-25.Truffle.TruffleGraal.GraalCompiler.GraalCompilerRoot.GraalCompiler.FrontEnd.HighTier.LoopUnswitchingPhase.InterceptException
                Exception occurred in scope: TruffleCompilerThread-25.Truffle.TruffleGraal.GraalCompiler.GraalCompilerRoot.GraalCompiler.FrontEnd.HighTier.LoopUnswitchingPhase.InterceptException
                Context obj org.graalvm.compiler.graph.VerificationError: is null input must not be null
                	at node: 3491|IsNull
                	at graph: StructuredGraph:33372{Method Vacation>>#initializeManager:	@330eef1, HotSpotMethod<OptimizedCallTarget.callRoot(Object[])>}
                Context obj org.graalvm.compiler.loop.phases.LoopUnswitchingPhase@59bab6c1
                Context obj org.graalvm.compiler.core.phases.HighTier@3f915ca3
                Context obj StructuredGraph:33372{Method Vacation>>#initializeManager:	@330eef1, HotSpotMethod<OptimizedCallTarget.callRoot(Object[])>}
                Context obj jdk.vm.ci.hotspot.HotSpotCodeCacheProvider@698baf6d
                Context obj Truffle<Method_Vacation>>#initializeManager:	@330eef1()>
[truffle] opt fail         Method Vacation>>#initializeManager:	@330eef1               |Reason org.graalvm.compiler.graph.VerificationError: is null input must not be null
	at node: 3491|IsNull
	at graph: StructuredGraph:33372{Method Vacation>>#initializeManager:	@330eef1, HotSpotMethod<OptimizedCallTarget.callRoot(Object[])>}
org.graalvm.compiler.graph.VerificationError: is null input must not be null
	at node: 3491|IsNull
	at graph: StructuredGraph:33372{Method Vacation>>#initializeManager:	@330eef1, HotSpotMethod<OptimizedCallTarget.callRoot(Object[])>}
	at org.graalvm.compiler.graph.Node.fail(Node.java:958)
	at org.graalvm.compiler.graph.Node.assertTrue(Node.java:945)
	at org.graalvm.compiler.nodes.calc.IsNullNode.verify(IsNullNode.java:81)
	at org.graalvm.compiler.graph.Graph.verify(Graph.java:1018)
	at org.graalvm.compiler.phases.BasePhase.apply(BasePhase.java:173)
	at org.graalvm.compiler.phases.BasePhase.apply(BasePhase.java:135)
	at org.graalvm.compiler.phases.PhaseSuite.run(PhaseSuite.java:154)
	at org.graalvm.compiler.phases.BasePhase.apply(BasePhase.java:153)
	at org.graalvm.compiler.phases.BasePhase.apply(BasePhase.java:135)
	at org.graalvm.compiler.core.GraalCompiler.emitFrontEnd(GraalCompiler.java:204)
	at org.graalvm.compiler.core.GraalCompiler.compile(GraalCompiler.java:177)
	at org.graalvm.compiler.core.GraalCompiler.compileGraph(GraalCompiler.java:163)
```

Unfortunately, the error does not occur with background compilation disabled.

To reproduce the problem, the following steps are needed:

```bash
git clone https://github.com/smarr/SOMns
cd SOMns
ant

bash -e ## just to stop on first error
export JVMCI_BIN=$YOUR_PATH_TO_YOUR_GRAAL_JVMCI_BINARY
export GRAAL_FLAGS=$ANY_GRAAL_FLAGS_PERHAPS
for i in {1..20}; do ./som -t1  core-lib/Benchmarks/Harness.som Vacation  150 0 14; done
```

In case you want the plain commandline of the invocation, use the following command:

```bash
./som -t1 -vv  core-lib/Benchmarks/Harness.som Vacation  150 0 14
```

**Comment by zapster** [comment]
Can it be reproduced with `-Dgraal.DumpOnError=true` and `-Dgraal.PrintIdealGraphFile=true`? A graph might help finding the issue.

**Comment by dougxc** [comment]
Also, can you try bisecting around recent loop related changes:
```
git log master --merges --pretty='format:%h %ad %>(20) %an %s' --date=format:'%d.%m.%y %H:%M' --grep=[lL]oop -n 5
```

**Comment by smarr** [comment]
With dump on error enabled, I get the following compilations*.cfg file:
```
begin_compilation
  name " Truffle::Method_Vacation>>#initializeManager:	@317ad1fc"
  method "unresolved Truffle Truffle.Method_Vacation>>#initializeManager:	@317ad1fc()"
  date 1484828663152
end_compilation
```

And the attached runtime graph:

[runtime-graphs-1484828663147_1.bgv.zip](https://github.com/graalvm/graal-core/files/716573/runtime-graphs-1484828663147_1.bgv.zip)

**Comment by smarr** [comment]
@dougxc Since I am using the version of Jan 14, the related commits in your list would be (or earlier).

```
cebca3f 09.01.17 03:01      Christian Wimmer [fix] Spill store optimization must not move stores into loop
35cc2e8 13.12.16 05:17    David Leopoldseder [fix] loop optimizations must respect ControlFlowAnchored interface
```

The version I used before was 98bdc0d327034162dd9aa59ef73ec1737b25aae0, which is from Dec. 8th.

I wont get to verifying which commit is more likely introducing it today unfortunately.

**Comment by tkrodriguez** [comment]
According to the IGV dump that IsNull node has no users so it's dead but hasn't been removed from the graph.  @gilles-duboscq recently made changes to killCFG so it's possible it's related to that somehow, so you might try the merge prior to that.

**Comment by tkrodriguez** [comment]
I reproduced this locally and the problem is that DominatorConditionalElimination leaves an unused floating node with valid inputs and then LoopUnswitchingPhase kills the Phi so the input becomes null.  I think IterativeConditionalEliminationPhase should be cleaning up nodes that become dead.  Currently it just ignores them.  That appears to fix this problem.  I'll prepare a changeset.

**Comment by smarr** [comment]
@tkrodriguez sorry for not getting back to this earlier. Just updated Graal, but can't yet confirm whether it is fixed, the same benchmark is hit by a new issue #271.

**Comment by tkrodriguez** [comment]
So this should have been fixed by 922be31 and the fix for #271 will merge in a bit and we can confirm them both.

**Comment by smarr** [comment]
Seems fixed. At least it doesn't reproduce at the moment. Thanks!
